### PR TITLE
Support for changes from draft-04

### PR DESCRIPTION
**Pull request by nicorsm**
## What's new?

Added support for new/changed fields from draft-04.
- renamed `group_state` to `context` in `HkdfLabel`
- renamed `transcript` to `transcript_hash` in `Group`
- added `supported_versions` to `UserInitKey`
- added `version` to `Welcome`
- added `index` to `Add`
- renamed `ApplicationPlaintext` to `ApplicationMessageContent`

## To-do
- Fix `test_uik_interop`  in `keys.rs`

## FYI
- Changes on the protocol from draft-03 to draft-04 are here: https://github.com/mlswg/mls-protocol/compare/draft-ietf-mls-protocol-03...draft-ietf-mls-protocol-04

### CODE REVIEW

**Comment by mikeger** [comment]
Thumbs up!

**Comment by nicorsm** [comment]
> Thumbs up!

Thanks @mikeger!! 😍😍

File: src/group.rs [hunk 1]
```diff
@@ -66,7 +66,7 @@ pub struct Group {
     roster: Vec<BasicCredential>,
     tree: Tree,
     update_secret: Option<(u64, NodeSecret)>,
-    transcript: Vec<GroupOperationValue>,
+    transcript_hash: Vec<GroupOperationValue>,
```
Comment (raphaelrobert) [inline src/group.rs]: The field type would also have to be a hash instead of a `Vec`, otherwise it's a bit confusing.

File: src/keys.rs [hunk 2]
```diff
@@ -489,15 +500,14 @@ fn test_user_init_key() {
     let mut buffer = Vec::new();
     uik.encode(&mut buffer);
 
-    let uik_hex = "020001002200203CB3FC6B9271B308EFEDC029502278DED42FC4AF181A44E31549F53B9BF7436C080700205F334D034259E2D6670D6CA8F5A937EA7CE9438259292F8872AEA6C7BB8AA2C000407DF11F6392DC7F1BD6FAFB34AA220C5457D2E58A2BB2C21DA4878A3E8AB8B0BA2AF2A87E7102D23DE169F880E38688406B34E582B6E978867755E37FB352DB0C";
+    let uik_hex = "020001002200203CB3FC6B9271B308EFEDC029502278DED42FC4AF181A44E31549F53B9BF7436C080700205F334D034259E2D6670D6CA8F5A937EA7CE9438259292F8872AEA6C7BB8AA2C002000100402A6B56527C2B8C942D0DD8CC5031663B7C43CE4D104C5CAF9A3EA270D4AF8F6DA75C5F68CB9E64CFCD10371ACF6E007F68F151C35162B474865159465E90A703";
```
Comment (raphaelrobert) [inline src/keys.rs]: The tests `test_user_init_key` and `test_uik_interop` are old and should be replaced by the binary test vectors. It's ok to delete these tests.

File: src/messages.rs [hunk 3]
```diff
@@ -176,9 +176,10 @@ pub struct Welcome {
     pub epoch: GroupEpoch,
     pub roster: Vec<BasicCredential>,
     pub tree: Vec<X25519PublicKey>,
-    pub transcript: Vec<GroupOperationValue>,
+    pub transcript_hash: Vec<GroupOperationValue>,
```
Comment (raphaelrobert) [inline src/messages.rs]: Same as mentioned above, we would need a real hash here of the actual transcript.

**Comment by nicorsm** [comment]
Thanks for the feedback @raphaelrobert!!

File: src/group.rs (hunk 1 above)
Comment (nicorsm) [reply src/group.rs]: @raphaelrobert  I have a doubt about the type... Protocol v4 says that _the `transcript_hash` field contains the list of `GroupOperation`_... shall we keep the complete list of `GroupOperation` values or just the last one by updating the hash with `transcript_hash_[n] = Hash(transcript_hash_[n-1] || operation)`?

**Comment by raphaelrobert** [comment]
Thanks for removing the tests @nicorsm , can you revert the changes to the transcript? This is not just something we can rename, we have to rewrite that part to be an actual hash.
I can then merge your PR.

**Comment by nicorsm** [comment]
@raphaelrobert done!

**Pull request merged**